serde unbacked bindings

avikchaudhuri opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #144894

Adds unbacked bindings during deserialization. These are carried by a node's metadata, and map pending fresh unbacked symbols to paths to such symbols inside the corresponding example value carried by the node's metadata. 

Since it is awkward to serialize paths, we only serialize the names of these symbols and reconstruct the paths on deserialization, using a shape env util. We also need to bump counters for unbacked symbols here, because the shape env util we use to create these symbols (when deserializing example values) don't do so, and not doing so makes later passes (like `run_decompositions`) crash because new unbacked symbols don't get new names.

This is enough for non-strict. For strict, the unbacked bindings and example values in node metadata can get out of sync, because of running AOTAutograd as an additional step after Dynamo. So we have to sync those back.

Differential Revision: [D68232274](https://our.internmc.facebook.com/intern/diff/D68232274/)

avikchaudhuri committed:
serde unbacked bindings

Adds unbacked bindings during deserialization for non-strict.

For strict, it seems that the unbacked bindings and example values in node metadata are out of sync, potentially because of running aot autograd as an additional step after Dynamo? Investigation is future work.

Differential Revision: [D68232274](https://our.internmc.facebook.com/intern/diff/D68232274/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "serde unbacked bindings"


Adds unbacked bindings during deserialization for non-strict.

For strict, it seems that the unbacked bindings and example values in node metadata are out of sync, potentially because of running aot autograd as an additional step after Dynamo? Investigation is future work.

Differential Revision: [D68232274](https://our.internmc.facebook.com/intern/diff/D68232274/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "serde unbacked bindings"


Adds unbacked bindings during deserialization.

This works directly for non-strict. For strict, the unbacked bindings and example values in node metadata can get out of sync, because of running AOTAutograd as an additional step after Dynamo. So we have to sync those back.

Differential Revision: [D68232274](https://our.internmc.facebook.com/intern/diff/D68232274/)

[ghstack-poisoned]

pianpwk commented on torch/_export/serde/serialize.py:
```diff
@@ -1839,6 +1845,20 @@ def _is_single_tensor_return(target) -> bool:
+                if symbol_is_type(u, SymT.UNBACKED_FLOAT):
+                    next(self.shape_env.unbacked_symfloat_counter)
+                elif symbol_is_type(u, SymT.UNBACKED_INT):
+                    next(self.shape_env.unbacked_symint_counter)
```
Hmm, so updating this counter now fixes existing run_decomp issues?

avikchaudhuri replied:
Without updating this counter, we will basically create new symbols with used names using the deserialized shape env. So, e.g., whenever we'd run `PropagateUnbackedSymInts`, as in `run_decompositions`, we might end up trying to `rebind_unbacked` the same symbol to itself, which is explicitly an assertion failure in that function.

We could solve this in a different way by re-implementing how we create symbols during deserialization. But I didn't want to do that here.

pianpwk commented on torch/export/_trace.py:
```diff
@@ -1391,6 +1391,22 @@ def _strict_export_lower_to_aten_ir(
+                node.meta["unbacked_bindings"] = new_unbacked_bindings
+            else:
+                del node.meta["unbacked_bindings"]
+
```
Sorry, what's the issue that's handled here? What I understand is AOTAutograd will regenerate new unbacked symbols with new & higher indices (which I guess is the out of sync here), and this pass is trying to re-index the symbols, to get the original symbols with lower indices? I thought this issue was resolved with unbacked_renamings?

avikchaudhuri replied:
The old symbols might now be equal to new symbols, or even ints, which will get returned any time we're looking inside the example values (by default, looking up an expression simplifies it). When serializing, we record symbols we see in the example values, so there's no record of the old symbols any more. Also, we don't record any renamings when serializing.

We could do this differently if we recorded the renamings. (But I didn't want to do that here.)

pianpwk replied:
ah, nice. I like this approach of keeping the original symbols.

pianpwk replied:
nit: maybe we can look into having this as part of the PropagateUnbackedSymInts interpreter (or subclass), it would be a bit cleaner and separated.

avikchaudhuri committed:
Update on "serde unbacked bindings"


Adds unbacked bindings during deserialization. These are carried by a node's metadata, and map pending fresh unbacked symbols to paths to such symbols inside the corresponding example value carried by the node's metadata. 

Since it is awkward to serialize paths, we only serialize the names of these symbols and reconstruct the paths on deserialization, using a shape env util. We also need to bump counters for unbacked symbols here, because the shape env util we use to create these symbols (when deserializing example values) don't do so, and not doing so makes later passes (like `run_decompositions`) crash because new unbacked symbols don't get new names.

This is enough for non-strict. For strict, the unbacked bindings and example values in node metadata can get out of sync, because of running AOTAutograd as an additional step after Dynamo. So we have to sync those back.

Differential Revision: [D68232274](https://our.internmc.facebook.com/intern/diff/D68232274/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "serde unbacked bindings"


Adds unbacked bindings during deserialization. These are carried by a node's metadata, and map pending fresh unbacked symbols to paths to such symbols inside the corresponding example value carried by the node's metadata. 

Since it is awkward to serialize paths, we only serialize the names of these symbols and reconstruct the paths on deserialization, using a shape env util. We also need to bump counters for unbacked symbols here, because the shape env util we use to create these symbols (when deserializing example values) don't do so, and not doing so makes later passes (like `run_decompositions`) crash because new unbacked symbols don't get new names.

This is enough for non-strict. For strict, the unbacked bindings and example values in node metadata can get out of sync, because of running AOTAutograd as an additional step after Dynamo. So we have to sync those back.

Differential Revision: [D68232274](https://our.internmc.facebook.com/intern/diff/D68232274/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "serde unbacked bindings"


Adds unbacked bindings during deserialization. These are carried by a node's metadata, and map pending fresh unbacked symbols to paths to such symbols inside the corresponding example value carried by the node's metadata. 

Since it is awkward to serialize paths, we only serialize the names of these symbols and reconstruct the paths on deserialization, using a shape env util. We also need to bump counters for unbacked symbols here, because the shape env util we use to create these symbols (when deserializing example values) don't do so, and not doing so makes later passes (like `run_decompositions`) crash because new unbacked symbols don't get new names.

This is enough for non-strict. For strict, the unbacked bindings and example values in node metadata can get out of sync, because of running AOTAutograd as an additional step after Dynamo. So we have to sync those back.

Differential Revision: [D68232274](https://our.internmc.facebook.com/intern/diff/D68232274/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "serde unbacked bindings"


Adds unbacked bindings during deserialization. These are carried by a node's metadata, and map pending fresh unbacked symbols to paths to such symbols inside the corresponding example value carried by the node's metadata. 

Since it is awkward to serialize paths, we only serialize the names of these symbols and reconstruct the paths on deserialization, using a shape env util. We also need to bump counters for unbacked symbols here, because the shape env util we use to create these symbols (when deserializing example values) don't do so, and not doing so makes later passes (like `run_decompositions`) crash because new unbacked symbols don't get new names.

This is enough for non-strict. For strict, the unbacked bindings and example values in node metadata can get out of sync, because of running AOTAutograd as an additional step after Dynamo. So we have to sync those back.

Differential Revision: [D68232274](https://our.internmc.facebook.com/intern/diff/D68232274/)

[ghstack-poisoned]
